interface/ceed-register.c

```diff
@@ -28,11 +28,15 @@ static bool register_all_called;
   @ref User
 **/
 int CeedRegisterAll() {
-  if (register_all_called) return 0;
-  register_all_called = true;
-
-#define CEED_BACKEND(name, ...) CeedChk(name());
+  int ierr = 0;
+  CeedPragmaCritical(CeedRegisterAll) {
+    if (!register_all_called) {
+      CeedDebugEnv256(1, "\n---------- Registering Backends ----------\n");
+#define CEED_BACKEND(name, ...) ierr += name();
 #include "../backends/ceed-backend-list.h"
 #undef CEED_BACKEND
-  return CEED_ERROR_SUCCESS;
+      register_all_called = true;
+    }
+  }
+  return ierr;
```

Comment:
Should we raise an error with a useful message here, not just return a count as though it were an enum value. Or are we not able to do so because error handling hasn't been defined yet?

Reply:
I just kept the behavior as before which used `CeedChk` on the return value (which has the behavior to just return `ierr` if nonzero and not to raise any error). It looks like the `ierr` value is handled by the caller in `CeedInit`, but there too it just uses `CeedCall` so the nonzero error code is returned to user and no errors are raised. I assume this behavior was an intentional choice at some point so would probably keep it for the sake of this PR unless you would rather address that change.

Reply:
So we probably should have made the error code strongly typed all along (so each function would return `CeedErrorType` rather than `int`), but the numbers have semantic meaning (`CeedErrorType`). We don't want that loop to do something like calculate `CEED_ERROR_MAJOR + CEED_ERROR_BACKEND == CEED_ERROR_UNSUPPORTED`. It's perhaps a side-show in this scenario, but it's confusing to have this pattern in the code. My preference would be to break early from the critical section so we return after the first error with the correct error code rather than charging ahead (possibly creating many errors) and returning what (the first or last or some mashup).

Reply:
How is [5230104](https://github.com/CEED/libCEED/pull/1279/commits/523010420027eb90f18075bec7b03c33137a94b3)? This should do exactly that rather than accumulating into `ierr` and changing the result. As soon as one registration fails, the rest are skipped in the critical section.